lib/ObjWriter/objwriter.cpp

```diff
@@ -596,31 +682,42 @@ extern "C" void EmitDebugFunctionInfo(ObjectWriter *OW,
   EmitLabelDiff(OST, SymbolsBegin, SymbolsEnd);
   OST.EmitLabel(SymbolsBegin);
   {
-    MCSymbol *ProcSegmentBegin = OutContext.createTempSymbol(),
-             *ProcSegmentEnd = OutContext.createTempSymbol();
-    EmitLabelDiff(OST, ProcSegmentBegin, ProcSegmentEnd, 2);
-    OST.EmitLabel(ProcSegmentBegin);
-
-    OST.EmitIntValue(unsigned(SymbolRecordKind::S_GPROC32_ID), 2);
-    // Some bytes of this segment don't seem to be required for basic debugging,
-    // so just fill them with zeroes.
-    OST.EmitFill(12, 0);
-    // This is the important bit that tells the debugger where the function
-    // code is located and what's its size:
-    OST.EmitIntValue(FunctionSize, 4);
-    OST.EmitFill(12, 0);
+    SymRecord RecHeader = {ulittle16_t(sizeof(ulittle16_t) + sizeof(ProcSym) +
+                                       strlen(FunctionName) + 1),
+                           ulittle16_t(SymbolRecordKind::S_GPROC32_ID)};
+    ProcSym ProRec = {};
+
+    ProRec.CodeSize = FunctionSize;
+    ProRec.DbgEnd = FunctionSize;
+
+    OST.EmitBytes(StringRef((char *)&RecHeader, sizeof(RecHeader)));
+    OST.EmitBytes(StringRef((char *)&ProRec, offsetof(ProcSym, CodeOffset)));
+
+    // Emit relocation
     OST.EmitCOFFSecRel32(Fn);
     OST.EmitCOFFSectionIndex(Fn);
-    OST.EmitIntValue(0, 1);
+
+    // Emit flags
+    OST.EmitIntValue(ProRec.Flags, sizeof(ProRec.Flags));
+
     // Emit the function display name as a null-terminated string.
-    OST.EmitBytes(FunctionName);
-    OST.EmitIntValue(0, 1);
-    OST.EmitLabel(ProcSegmentEnd);
+    OST.EmitBytes(StringRef(FunctionName, strlen(FunctionName) + 1));
+
+    // Emit local var info
+    int NumVarInfos = OW->DebugVarInfos.size();
+    if (NumVarInfos > 0) {
+      std::unique_ptr<DebugVarInfo[]> varInfos(new DebugVarInfo[NumVarInfos]);
+      std::copy(OW->DebugVarInfos.begin(), OW->DebugVarInfos.end(),
```

Comment:
Do you really need a copy here since you've already retained the data structure in vector?
You could simply pass &OW->DebugVarInfos[0] to EmitCVDebugVarInfo instead of the copied on varInfos.get().